include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -0,0 +1,1043 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 - 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_differentiation_ad_ad_number_traits_h
+#define dealii_differentiation_ad_ad_number_traits_h
+
+#include <deal.II/base/numbers.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/differentiation/ad/ad_number_types.h>
+
+#include <boost/type_traits.hpp>
+
+#include <complex>
+#include <type_traits>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Differentiation
+{
+  namespace AD
+  {
+
+    /**
+     * A number traits class to help describe some characteristic
+     * information about auto-differentiable numbers.
+     *
+     * @tparam ScalarType A real or complex floating point number.
+     * @tparam ADNumberTypeCode An enumeration specifying the type
+     *         code for the supported auto-differentiable counterpart
+     *         to the given @p ScalarType.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template<typename ScalarType, enum NumberTypes ADNumberTypeCode, typename = void>
+    struct NumberTraits;
+
+
+
+    /**
+     * A number traits class to help describe some characteristic
+     * information about auto-differentiable numbers. This class is only
+     * specialized (and, therefore, only made constructible) if the
+     * @p ADNumberType template parameter is indeed a supported
+     * auto-differentiable number.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template<typename ADNumberType, typename = void>
+    struct ADNumberTraits;
+
+
+    /**
+     * This namespace defines the classes that help provide a unified interface
+     * to all auto-differentiation numbers.
```

Comment:
...numbers in Adol-C?

Reply:
And Sacado (and whatever other libraries someone might want to add in the future).

Reply:
ah, yes, i think i somehow was under impression that this part is in adol-c namespace, but it clearly is not.